src/systems/a320_systems/src/hydraulic.rs

```diff
@@ -1479,28 +1547,217 @@ impl SimulationElement for A320BrakingForce {
+    control_position_request: Ratio,
+    should_unlock: bool,
 }
-impl Door {
-    fn new(id: usize) -> Self {
+impl A320DoorController {
+    // Duration which the hydraulic valves sends a open request when request is closing (this is done on real aircraft so uplock can be easily unlocked without friction)
+    const UP_CONTROL_TIME_BEFORE_DOWN_CONTROL: Duration = Duration::from_millis(200);
+
+    // Delay from the ground crew unlocking the door to the time they start requiring up movement in control panel
+    const DELAY_UNLOCK_TO_HYDRAULIC_CONTROL: Duration = Duration::from_secs(5);
+
+    fn new(id: &str) -> Self {
         Self {
-            exit_id: format!("EXIT OPEN:{}", id),
-            position: 0.,
-            previous_position: 0.,
+            requested_position_id: format!("{}_DOOR_CARGO_OPEN_REQ", id),
+            state: DoorControlState::DownLocked,
+            position_requested: Ratio::new::<ratio>(0.),
+
+            duration_in_no_control: Duration::from_secs(0),
+            duration_in_hyd_control: Duration::from_secs(0),
+
+            should_close_valves: true,
+            control_position_request: Ratio::new::<ratio>(0.),
+            should_unlock: false,
         }
     }
 
-    fn has_moved(&self) -> bool {
-        (self.position - self.previous_position).abs() > f64::EPSILON
+    fn update(&mut self, context: &UpdateContext, door: &CargoDoor, pressure_available: Pressure) {
+        self.state = self.determine_mode(door, pressure_available);
+        self.update_timers(context);
+        self.update_actions_from_state();
+    }
+
+    fn update_timers(&mut self, context: &UpdateContext) {
+        if self.state == DoorControlState::NoControl {
+            self.duration_in_no_control += context.delta();
+        } else {
+            self.duration_in_no_control = Duration::from_secs(0);
+        }
+
+        if self.state == DoorControlState::HydControl {
+            self.duration_in_hyd_control += context.delta();
+        } else {
+            self.duration_in_hyd_control = Duration::from_secs(0);
+        }
+    }
+
+    fn update_actions_from_state(&mut self) {
+        match self.state {
+            DoorControlState::DownLocked => {}
+            DoorControlState::NoControl => {
+                self.should_close_valves = true;
+            }
+            DoorControlState::HydControl => {
+                self.should_close_valves = false;
+                if self.position_requested > Ratio::new::<ratio>(0.)
+                    || self.duration_in_hyd_control < Self::UP_CONTROL_TIME_BEFORE_DOWN_CONTROL
+                {
+                    self.control_position_request = Ratio::new::<ratio>(1.);
+                } else {
+                    self.control_position_request = Ratio::new::<ratio>(0.);
+                }
+            }
+            DoorControlState::UpLocked => {
+                self.should_close_valves = true;
+            }
+        }
+    }
+
+    fn determine_mode(
```

Comment:
**thought**: `determine_mode` returns `DoorControlState`. Should `determine_mode` be named `determine_state`?
**thought**: The function also contains writes to `self.should_unlock`, thus `determine_state` might not fully capture the intent either.